File: smoke_test.py
"""
Pre-flight smoke test.

Runs a minimal end-to-end check against the sandbox:
    1. Load config — verifies all required env vars are present.
    2. Tiny Search call — verifies Client (or Search) token works.
    3. Tiny Chat call — verifies Client token + Chat scope.
    4. Tiny Indexing dry-run — verifies Indexing token (does not actually index).

Run this:
    - After first-time setup.
    - Before starting any work session, to catch token rotation / env issues.
    - Before the live interview. Twice.

Exit code 0 on success, 1 on any failure.
"""
from __future__ import annotations

import sys
import traceback

from src.config import load_config
from src.glean_client import IndexingClient, QueryClient


def _banner(msg: str) -> None:
    print(f"\n{'=' * 60}\n  {msg}\n{'=' * 60}")


def main() -> int:
    failures: list[str] = []

    _banner("1. Load config")
    try:
        cfg = load_config()
        print(f"   instance   : {cfg.glean_instance}")
        print(f"   datasource : {cfg.glean_datasource}")
        print(f"   indexing tok: ...{cfg.glean_indexing_token[-6:]}")
        print(f"   client tok  : ...{cfg.glean_client_token[-6:]}")
        print(f"   search tok  : "
              f"{'...' + cfg.glean_search_token[-6:] if cfg.glean_search_token else '(using client token)'}")
    except Exception as e:
        print(f"   FAIL: {e}")
        return 1

    _banner("2. Search API (Client/Search token)")
    try:
        q = QueryClient(cfg)
        result = q.search(query="test", page_size=1, datasource=cfg.glean_datasource)
        got = len(result.json.get("results", []))
        print(f"   OK — {result.status} in {result.latency_ms}ms, {got} result(s)")
    except Exception as e:
        failures.append(f"search: {e}")
        print(f"   FAIL: {e}")
        traceback.print_exc()

    _banner("3. Chat API (Client token)")
    try:
        q = QueryClient(cfg)
        result = q.chat(message="Say 'ok' if you can read this.", context_docs=[])
        print(f"   OK — {result.status} in {result.latency_ms}ms")
    except Exception as e:
        failures.append(f"chat: {e}")
        print(f"   FAIL: {e}")
        traceback.print_exc()

    _banner("4. Indexing API auth check (Indexing token)")
    # We do NOT actually push a doc here — we construct the client and let the
    # caller decide whether to exercise it. Token validity is implicitly
    # checked the first time the indexer runs. If you want a stronger check,
    # uncomment the index_documents call below with a synthetic doc.
    try:
        _ = IndexingClient(cfg)
        print("   OK — IndexingClient constructed; token format accepted.")
        print("   (Run `python -m src.indexer` to exercise the full path.)")
    except Exception as e:
        failures.append(f"indexing client: {e}")
        print(f"   FAIL: {e}")

    _banner("Summary")
    if failures:
        print(f"FAILED ({len(failures)}):")
        for f in failures:
            print(f"  - {f}")
        return 1
    print("All checks passed. Environment is ready.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: server/scripts/smoke-test.ts
/**
 * Pre-flight smoke test.
 *
 * Runs a minimal end-to-end check against the sandbox:
 *     1. Load config — verifies all required env vars are present.
 *     2. Tiny Search call — verifies Client (or Search) token works.
 *     3. Tiny Chat call — verifies Client token + Chat scope.
 *     4. Tiny Indexing dry-run — verifies Indexing token (does not actually index).
 *
 * Run this:
 *     - After first-time setup.
 *     - Before starting any work session, to catch token rotation / env issues.
 *     - Before the live interview. Twice.
 *
 * Exit code 0 on success, 1 on any failure.
 */

import { loadConfig } from '../config.js';
import { IndexingClient, QueryClient } from '../glean-client.js';

function banner(msg: string): void {
  console.log(`\n${'='.repeat(60)}\n  ${msg}\n${'='.repeat(60)}`);
}

function maskToken(token: string): string {
  if (!token || token.length < 6) return '(invalid)';
  return `...${token.slice(-6)}`;
}

async function smokeTest(): Promise<number> {
  const failures: string[] = [];

  banner('1. Load config');
  let cfg;
  try {
    cfg = loadConfig();
    console.log(`   instance    : ${cfg.config.gleanInstance}`);
    console.log(`   datasource  : ${cfg.config.gleanDatasource}`);
    console.log(`   indexing tok: ${maskToken(cfg.config.gleanIndexingToken)}`);
    console.log(`   client tok  : ${maskToken(cfg.config.gleanClientToken)}`);
    console.log(`   search tok  : ${cfg.config.gleanSearchToken ? maskToken(cfg.config.gleanSearchToken) : '(using client token)'}`);
  } catch (error) {
    console.log(`   FAIL: ${error instanceof Error ? error.message : error}`);
    return 1;
  }

  banner('2. Search API (Client/Search token)');
  try {
    const queryClient = new QueryClient(cfg);
    const result = await queryClient.search('test', 1, cfg.config.gleanDatasource);
    const resultCount = (result.json as any).results?.length || 0;
    console.log(`   OK — ${result.status} in ${result.latencyMs}ms, ${resultCount} result(s)`);
  } catch (error) {
    failures.push(`search: ${error instanceof Error ? error.message : error}`);
    console.log(`   FAIL: ${error instanceof Error ? error.message : error}`);
    if (error instanceof Error && error.stack) {
      console.log(error.stack);
    }
  }

  banner('3. Chat API (Client token)');
  try {
    const queryClient = new QueryClient(cfg);
    const result = await queryClient.chat("Say 'ok' if you can read this.", []);
    console.log(`   OK — ${result.status} in ${result.latencyMs}ms`);
  } catch (error) {
    failures.push(`chat: ${error instanceof Error ? error.message : error}`);
    console.log(`   FAIL: ${error instanceof Error ? error.message : error}`);
    if (error instanceof Error && error.stack) {
      console.log(error.stack);
    }
  }

  banner('4. Indexing API auth check (Indexing token)');
  // We do NOT actually push a doc here — we construct the client and let the
  // caller decide whether to exercise it. Token validity is implicitly
  // checked the first time the indexer runs. If you want a stronger check,
  // uncomment the index_documents call below with a synthetic doc.
  try {
    const indexingClient = new IndexingClient(cfg);
    console.log('   OK — IndexingClient constructed; token format accepted.');
    console.log('   (Run `npm run index` to exercise the full path.)');
  } catch (error) {
    failures.push(`indexing client: ${error instanceof Error ? error.message : error}`);
    console.log(`   FAIL: ${error instanceof Error ? error.message : error}`);
  }

  banner('Summary');
  if (failures.length > 0) {
    console.log(`FAILED (${failures.length}):`);
    failures.forEach((f) => console.log(`  - ${f}`));
    return 1;
  }
  console.log('All checks passed. Environment is ready.');
  console.log('\nNext steps:');
  console.log('  1. Run: npm run index');
  console.log('  2. Run: npm run dev');
  console.log('  3. Open: http://localhost:5173');
  return 0;
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  smokeTest()
    .then((exitCode) => {
      process.exit(exitCode);
    })
    .catch((error) => {
      console.error('\n❌ Smoke test crashed:', error);
      process.exit(1);
    });
}

export { smokeTest };
